dialer.go

```diff
@@ -226,6 +230,39 @@ func (d *Dialer) connectTLS(ctx context.Context, conn net.Conn) (tlsConn *tls.Co
 	return
 }
 
+func (d *Dialer) authenticateSASL(ctx context.Context, conn net.Conn) error {
+	// Temporary create a new conn.
+	// TODO: Can we move authentication to a later step when Conn is available?  Is it an issue that we might reuse correlation ids?
```

Comment:
In my code I did the authentication on `Conn` instead of `net.Conn`. The changes should be fairly straight-forward, change this function to take a `Conn` instead of a `net.Conn` and call it from `DialContext` and `DialLeader`.

To me it feels a bit cleaner to not create a new `Conn` just for the authentication.